### PR TITLE
feat(ir): implement `ArrayMap` operation

### PR DESCRIPTION
**Pull request by kszucs**
```py
a = ibis.array([1, 2, 3])
r = a.map(lambda x: x * 2.1)
assert r.type() == dt.Array(dt.float64)
```

Just a quick prototype for now, adding more thorough validation, testing and implementation for some backends.

cc @cpcloud @NickCrews

### CODE REVIEW

**Comment by cpcloud** [comment]
It'll be pretty easy to follow this up with `ArrayFilter` and perhaps others as the need arises.

**Comment by kszucs** [comment]
Can you hold off merging this a bit, I'm experimenting with a cleaner approach.

**Comment by cpcloud** [comment]
> Can you hold off merging this a bit, I'm experimenting with a cleaner approach.

Yep, moving to draft!

**Comment by cpcloud** [comment]
@kszucs Is this ready for review/merge?

**Review by kszucs — COMMENTED** [review]
+1

File: ibis/expr/operations/arrays.py [hunk 1]
```diff
@@ -88,6 +91,36 @@ class ArrayRepeat(Value):
...
+    # and return type, e.g. Callable[[ops.Argument], ops.Value] in this case
+    func = rlz.instance_of(Callable)
+
+    output_shape = rlz.shape_like("arg")
+
+    @attribute.default
+    def result(self):
+        shape = rlz.Shape.COLUMNAR
+        dtype = self.arg.output_dtype.value_type
+        args = [
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/expr/operations/arrays.py:108]: Updating in a follow-up #5554

File: ibis/expr/operations/arrays.py [hunk 2]
```diff
@@ -88,6 +91,36 @@ class ArrayRepeat(Value):
     output_shape = rlz.shape_like("args")
 
 
+@public
+class ArrayMap(Value):
+    arg = rlz.array
+    # TODO(kszucs): add a callable_with validator to support callable arguments
+    # and return type, e.g. Callable[[ops.Argument], ops.Value] in this case
+    func = rlz.instance_of(Callable)
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/expr/operations/arrays.py:99]: Added support for it in #5554

**Pull request merged**